### PR TITLE
Add internals to docs

### PR DESCRIPTION
**Pull request by rushk014**

### CODE REVIEW

**Review by jpdunc23 — COMMENTED** [review]
Looks good. Did some revision. I'm going to have Sahil take a look so we can see if this is clear to a new user.

**Comment by jpdunc23** [comment]
> Looks good. Did some revision. I'm going to have Sahil take a look so we can see if this is clear to a new user.

@ssaxena00 hope you can read this from the perspective of a new user and let us know what's not clear.

File: docs/internals.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+## Vset
+
+Fundamentally, all `vflow.vset.Vset` objects generate permutations by applying
+their modules to the Cartesian product of input parameters. Internally, `Vset`
+modules and parameters are wrapped in dictionaries with tuple keys. These tuples
+contain `vflow.subkey.Subkey` objects, which identify the `vflow.vfunc.Vfunc` or
+parameter name (the `Subkey` "value"), as well as its originating `Vset` (the
+`Subkey` "origin"). Since inputs must be dictionaries with a certain format, all
+raw inputs must be initialized using `vflow.helpers.init_args`.
```
Comment (ssaxena00) [inline docs/internals.md:18]: A little confused by the description, might be helpful to show what the dictionary looks like. This is my understanding, not sure what the values in the dictionaries are:
{ (`Subkey` "value", `Subkey` "origin") : ___ }

File: docs/internals.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+regardless of when it was called, as with a data cleaning `Vset` that is used
+first on a training dataset before model fitting and later on a test dataset for
+model evaluation; this ensures that the data cleaning procedure used in training
+the model matches the data cleaning procedure used when evaluating the model.
+(In contrast, note that in this example we wouldn't need to initialize the
+modeling `Vset` with `output_matching=True`.) On the other hand, if separate
+calls to the `Vset` result in entirely independent outputs, such as is usually
+the case for a `Vset` that does subsampling, then the `Vset` should be
+initialized with `output_matching=False` (the default) to prevent bad matches
+(e.g., unnecessary matching on subsamples of two different datasets).
```
Comment (ssaxena00) [inline docs/internals.md:48]: This paragraph makes sense, but feels somewhat overloaded. A suggestion could be creating a list like below:

When to use `output_matching=True`:
- etc
- etc

When to use `output_matching=False`:
- etc
- etc

File: docs/internals.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+## Vfunc
+
+When a `Vset` is initialized, the items in the `modules` arg are wrapped in
+`Vfunc` objects, which are like named functions that can optionally support a
+`fit` or `transform` method. If a `Vset` is initialized with `is_async=True`,
+then modules are wrapped in `AsyncModule` objects, which compute asynchronously
+using `ray`. If a `Vset` is initialized with `lazy=True`, `Vset` output
+dictionary values are wrapped in `VfuncPromise` objects, which are lazily
+evaluated. At the moment, `VfuncPromise` objects are only resolved if passed to
+a `Vset` with `lazy=False` downstream or if called manually by the user.
```
Comment (ssaxena00) [inline docs/internals.md:59]: Similarly here, seems like an easier to read format could be writing a short description and then bullet points like: 

When a `Vset` is initialized, the items in the `modules` arg are wrapped in `Vfunc` objects, which are like named functions that can optionally support a `fit` or `transform` method. 

- `is_async=True`: modules are wrapped in `AsyncModule` objects, which compute asynchronously
using `ray`
- `lazy=True`: `Vset` output dictionary values are wrapped in `VfuncPromise` objects, which are lazily evaluated. At the moment, `VfuncPromise` objects are only resolved if passed to a `Vset` with `lazy=False` downstream or if called manually by the user.

File: docs/internals.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+tuple keys containing a `Subkey` with origin "cleaning" and value "median
+impute", and that `Subkey` is carried through when the cleaned data is passed to
+a modeling `Vset`. When we want to produce test predictions from the modeling
+`Vset`, `vflow` decides which trained models to apply to which version of the
+cleaned test data based on the `Subkey`s created by the data cleaning `Vset`. We
+have a positive match if the tuple keys each have a `Subkey` with origin
+"cleaning" and value "median impute". On the other hand, if one of the `Subkey`s
+instead had value "mean impute", then the two keys mismatch and we proceed to
+check the next dictionary key without applying the trained model to the test
+data.
```
Comment (ssaxena00) [inline docs/internals.md:93]: Might be easier to follow by using subheadings in this paragraph. Could be something like:

#### When `vflow` combines two dictionaries it previously created
...

#### When the `Vset` is applied to the training data
...

#### When the modeling `Vset` produces test predictions
...

**Review by ssaxena00 — APPROVED** [review]

**Pull request merged**